_posts/2020-04-07-how-to-start-with-opentelemetry-in-nodejs.markdown

```diff
@@ -0,0 +1,300 @@
+**Discaimer**: I can not define myself as a NodeJS developer. I wrote a bunch of
+AngularJS single page application back in the day, I wrote some Cordova mobile
+applications ages ago. I am not writing any JS production code since 2015 more
+or less.
+
+## First approach
+
+I concluded the application instrumentation it was the day the maintainers
+tagged the first beta release. Overnight I had to update libraries and test
+code. Very luckily.
+
+The way I learned about how to properly instrument
+[discount](https://github.com/gianarb/shopmany/tree/master/discount) required a
+lot of digging in the actual
+[opentelemery-js](https://github.com/open-telemetry/opentelemetry-js) but
+luckily for us it has a lot of examples and the library is designed to load a
+bunch of useful modules that are able to instrument the application by itself.
+The community is very helpful and you can chat via
+[Gitter](http://gitter.im/open-telemetry/opentelemetry-js).
+
+## Getting Started
+
+I am using ExpressJS and OpenTelemetry has a plugin for it that you can load,
+and it instruments the app by itself, same for MongoDB that is the packaged I am
+using.
+
+Those are the dependencies I installed in my applications, all of them are
+provided by the repository I linked above:
+
+```
+"@opentelemetry/api": "^0.5.0",
+"@opentelemetry/exporter-jaeger": "^0.5.0",
+"@opentelemetry/node": "^0.5.0",
+"@opentelemetry/plugin-http": "^0.5.0",
+"@opentelemetry/plugin-dns": "^0.5.0",
+"@opentelemetry/plugin-mongodb": "^0.5.0",
+"@opentelemetry/tracing": "^0.5.0",
+"@opentelemetry/plugin-express": "^0.5.0",
+```
+
+I created a `./tracer.js` file that initialize the tracer, I have added inline
+documentation to explain the crucial part of it:
+
+```
+js'use strict';
+
+const opentelemetry = require('@opentelemetry/api');
+const { NodeTracerProvider } = require('@opentelemetry/node');
+const { SimpleSpanProcessor } = require('@opentelemetry/tracing');
+// I am using Jaeger as exporter
+const { JaegerExporter } = require('@opentelemetry/exporter-jaeger');
+
+// This is not mandatory, by default httptrace propagation is used
+// but it is not well supported by the PHP ecosystem and I have
+// a PHP service to instrument. I discovered B3 is supported
+// form all the languages I where intrumenting
+const { B3Propagator } = require('@opentelemetry/core');
+
+module.exports = (serviceName, jaegerHost, logger) => {
+  // A lot of those plugins are automatically loaded when you install them
+  // So if you do not use express for example you do not have to enable all
+  // those plugins manually. But Express is not auto enabled so I had to add them
+  // all
+  const provider = new NodeTracerProvider({
+    plugins: {
+      dns: {
+        enabled: true,
+        path: '@opentelemetry/plugin-dns',
+      },
+      mongodb: {
+        enabled: true,
+        path: '@opentelemetry/plugin-mongodb',
+      },
+      http: {
+        enabled: true,
+        path: '@opentelemetry/plugin-http',
```

Comment:
You generally want to avoid tracing paths used by your service to verify its health like follow:
```
http: {
  enabled: true,
  path: '@opentelemetry/plugin-http',
  ignoreIncomingPaths: [
    '/',
    '/health'
  ]
}
```